### ISSUE TITLE
torch.autograd.grad support 

### ISSUE THREAD

**Issue by cdancette** [f-dangel/backpack#155]
Hi,
Thanks for this great library.
It seems that the `torch.autograd.grad` function is not supported with backpack, are you planning to add support ?

It would be useful to compute batch grads with respect to intermediate features for example.

**Comment by f-dangel** [comment]
Hi,

apologize for the delayed reply, another project kept me busy over the last weeks. 

Now to your question: It really depends what you are trying to do. BackPACK was specifically designed to work with calls to `backward()`. But since it relies on running additional code through backward hooks whenever `torch.autograd.xxx` computes gradients, those hooks should be triggered, regardless whether you perform a backward pass with `torch.autograd.backward` or `torch.autograd.grad`.

Here is an MWE that computes individual gradients with BackPACK in a backward pass with the two different functions:

```python
"""
Test if BackPACK works with torch.autograd.grad.

Both torch.autograd.backward and torch.autograd.grad should trigger BackPACK's
backward hooks.
"""

import torch

from backpack import backpack, extend, extensions
from backpack.utils.examples import load_one_batch_mnist

torch.manual_seed(0)

N = 3
X, y = load_one_batch_mnist(batch_size=N)

model = torch.nn.Sequential(
    torch.nn.Flatten(),
    torch.nn.Linear(784, 64),
    torch.nn.ReLU(),
    torch.nn.Linear(64, 10),
)
lossfunc = torch.nn.CrossEntropyLoss()

model = extend(model)
lossfunc = extend(lossfunc)


def via_autograd_backward():
    """Compute individual gradients with BackPACK and autograd.backward."""
    loss = lossfunc(model(X), y)

    with backpack(extensions.BatchGrad(), debug=True):
        loss.backward()

    igrads = []

    for p in model.parameters():
        igrads.append(p.grad_batch)
        del p.grad_batch

    return igrads


def via_autograd_grad():
    """Compute individual gradients with BackPACK and autograd.grad."""
    loss = lossfunc(model(X), y)

    with backpack(extensions.BatchGrad(), debug=True):
        torch.autograd.grad(loss, list(model.parameters()))

    igrads = []

    for p in model.parameters():
        igrads.append(p.grad_batch)
        del p.grad_batch

    return igrads


igrad_autograd_grad = via_autograd_grad()
igrad_autograd_backward = via_autograd_backward()

assert len(igrad_autograd_grad) == len(igrad_autograd_backward)

for igrad1, igrad2 in zip(igrad_autograd_backward, igrad_autograd_grad):
    assert torch.allclose(igrad1, igrad2)
```

Do you have a specific example how one would combine `torch.autograd.grad` with BackPACK?

Regarding your question about computing gradients w.r.t. hidden features: Do you mean accessing the gradients that are backpropagated by PyTorch? If so, then you can just add a [backward hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=register_full#torch.nn.Module.register_full_backward_hook) to the layer whose intermediate gradients you are interested in.

**Comment by cdancette** [comment]
Thanks for your answer !

My use case for torch.autograd.grad is to compute gradients of the loss with respect to intermediate features. So I guess torch.autograd.grad is not the issue here, but the fact that backpack only stores batch gradients for module parameters, even if a tensor has the `retain_grad()` activated.

 Here is an example:

```python
import torch
from backpack import extend, backpack
from backpack.extensions import BatchGrad

X = torch.rand(5, 3)
y = torch.randint(0, 9, (5,))

model1 = torch.nn.Linear(3, 3)
model2 = torch.nn.Linear(3, 10)

features = model1(X)
features.retain_grad()
# features = extend(features)

loss = model2(features).max(1).values.sum()

with backpack(BatchGrad()):
    grad, = torch.autograd.grad(loss, inputs=features, create_graph=True, retain_graph=True)

print(grad.shape)
print(features.grad_batch.shape)
```

Here, the last line returns an error. I tried adding `features = extend(features)`, but it fails since it is a tensor and not a pytorch module. Do you think there is a way to make this work with backpack ?

**Comment by cdancette** [comment]
You are right, I already have per-samples gradients for my features. 

I oversimplified the thing I am trying to achieve. I actually not want per-sample grad, but per-dimension grad, where the dimension comes from having multiple models in parallel (one loss for each).  I think backpack is not designed for this setup.

But I think I can actually solve this with torch.autograd.functional.jacobian ! 

Thanks again for your answers.
